### ISSUE TITLE
Contact imports 2.0

### ISSUE THREAD

**Issue by rowanseymour** [rapidpro/rapidpro#1340]
**Goals:**

 1. Move database operations to mailroom for performance and reliability
 2. Imports should be robust to deploys - i.e. mailroom can be stopped and started and task will resume where it left off
 3. Provide users with visual feedback of status.. if not progress
 4. Support other possible pathways in the future for bulk creating contacts, e.g. user asked recently for a bulk create API endpoint ?
 5. Creating an architecture for long running tasks which is applicable to other things like exports ?

**Ideas:**

Unpacking 3.. I think the user

 * Sees something on the contact list page to say that an import is in progress
 * Sees contacts being added in realtime.. this is as much a side effect of not using a transaction as a feature
 * Can't start simultaneous imports

Unpacking 2. maybe we need to break the task down into smaller units - which can always finish in a reasonable amount of time. For example importing 50K contacts actually becomes 50 smaller tasks of importing 1K contacts.

I think it would help with 2. and make 4. possible to use a two step process where rows from the Excel file are parsed into a intermediate format before the database work is done..

1. Parse and validate the Excel file 
2. Create batches with the contacts described in the "contact spec" format used by create contact endpoint in mailroom
3. Run those batches 

Step 1. could still live in RapidPro and I think it being there makes it easier to do validation and send those errors back to the user, localised etc.
Step 2. could be in either
Step 3. definitely lives in Mailroom

**Comment by nicpottier** [comment]
Breaking up into chunks would in theory allow parallel importing, which should work just fine it seems like since they are contact specific. And that would be a big performance boost. That does mean that we are basically keeping the import in Redis memory though (and in an exploded JSON form no less) but I think that is probably ok for a short term thing.

Task retries in the event of failure shouldn't be a big deal and as you said as long as we aren't killing -9 mailroom, then if they are reasonably sized (maybe 500 contacts each) then they should complete reasonably quickly that a stop of mailroom should happen quickly without dropping any on the floor.

I think my bias would be for 1,2,3 to all be in mailroom just because whatever is checking the format should probably be the same thing that is doing the importing, otherwise say the Python library could decide it can read things fine but Golang blows up.

As for progress, if we break into 500 contact chunks then in theory we should be able to track how many chunks we have done and show that somewhere pretty easily. (can use a redis key to increment that until completion)

**Comment by rowanseymour** [comment]
The parallel thing could bite us tho if a lot of 500-contact chunks are running simultaneously then.. you might lose the guarantee that that they finish in a reasonable time.. then you do risk them getting dropped during mailroom shutdown.

> otherwise say the Python library could decide it can read things fine but Golang blows up.

To be clear I'm suggesting RP parses the Excel file, if there are problems those go back to UI, and if not, creates contact specs that we mailroom can already parse.. it's not like they would be reading the same thing twice in different code bases.

Of course there might be errors that happen after step 1 - for example immediately after starting the import the user deletes a contact field... but that could happen regardless of whether mailroom or RP is doing step 1. I actually think you just ignore those actually and create the contact without that field or group or whatever is found to be missing.

My stance on 1. is that I don't see a benefit in moving that to mailroom - it just makes it harder to get localized error messages back to the user etc. Also we know the Python libraries for XLS/XLSX files work...

But I think also we can approach this incrementally.. get 3. working first of all in mailroom.

**Comment by nicpottier** [comment]
> you might lose the guarantee that that they finish in a reasonable time.. then you do risk them getting dropped during mailroom shutdown.

Why do you say that? Because chunks may take longer to process and finish in parallel? True they might but I can't think of any reason why this wouldn't work pretty well in parallel at reasonable levels. I guess up to the # of CPUs on the DB.

> To be clear I'm suggesting RP parses the Excel file, if there are problems those go back to UI, and if not, creates contact specs that we mailroom can already parse.. it's not like they would be reading the same thing twice in different code bases.

Oh ok, so mailroom never sees Excel at all. I guess that's fine ya, as long as Celery isn't killed during that batch creation!

**Comment by rowanseymour** [comment]
> Why do you say that? 

🤷  Because I have little idea what the performance of hundreds of simultaneous import batches would be.. but yeah as long as we have a sensible upper limit on import size.. then it should be fine. Will be exciting to see what kind of a performance win we can get from parallelism.

**Comment by nicpottier** [comment]
Well hopefully we aren't doing hundreds at once.. we run mailroom with 2 batch workers per box..

**Comment by rowanseymour** [comment]
ah ok then nevermind!

**Comment by rowanseymour** [comment]
Ok I think this is what it looks like...

1. We have a new `ContactImport` model since we don't really get much value from using `csv_imports.ImportTask` and mostly fight against its lack of things like an org field.
2. As with existing code, there's a header validation stage and customize stage that allows mapping of fields.
3. Create task in RP to read the Excel file on that and parse into JSON contact specs. We're not writing anything to the DB so this should be relatively fast and doable in < 2 minutes (celery tasks get 2 minutes to finish during deploys). And we've also talked about having a limit of 50K rows on imports.
4. As we go, load those into Mailroom tasks in Redis - batches of 500
5. Mailroom task handles those asynchronously

One challenge perhaps is how mailroom communicates back to RP the status/progress of the import. I think it would be nice to show that there is an ongoing import in the contact list page UI (ideas @ericnewcomer ?). 

If `ContactImport` had a field for the number of batches handled, and the total number of batches, then when each batch in mailroom finishes, it can increment the former field, and that would give us value that could go into a progress bar in the RP UI.

**Comment by rowanseymour** [comment]
Alternatively `ContactImportBatch` exists as a DB thing. RP creates these and mailroom handles them and updates them. Postgres supposedly compresses all text fields so this model could also hold the contact specs to avoid writing them all to Redis.

The advantage of this is that each task in mailroom now has its own place to record its results and there's no danger of collisions. For example we need to record per-row errors. We can't have a single field `ContactImport` that each mailroom task writes to because they'll overwrite each other.

Displaying the results of the import in RapidPro is done by merging across the batch objects for that import. Maybe something like...

```
contactimport
  org_id int
  file text

contactimportbatch
  import_id int
  status queued|processing|complete
  payload json
  errors []text
  num_created int
  num_updated int
```

**Comment by ericnewcomer** [comment]
I would love to have access to an object that showed overall progress for an import. We perhaps could show the group before it is ready as a form of status.

**Comment by nicpottier** [comment]
I don't hate the batch approach, those can be trimmed obviously so there's no long term DB hit there. It will definitely slow down the creation process vs just blasting to Redis, but inserts should be pretty fast regardless. As you said it also makes tracking status a lot saner, same with errors. And since no two tasks should be touching the same batch we shouldn't run into locking issues.

So that gets a +1 from me. If we do batches of 500 contacts then that's only 50 rows for 25k contacts, so no biggie. (as an aside we should make that max size easily configurable, if it is faster than we expect then maybe we can do 100k instead)

It may be nicer to have errors be slightly more structured,  IE, a JSON blob so that we can call out line numbers and format those errors a bit better. May also be nice to have an elapsed on a batch to track our performance characteristics.

**Comment by rowanseymour** [comment]
Added in https://github.com/nyaruka/rapidpro/pull/2986
